### PR TITLE
FlightTasks: pass the timestamp of the message that causes FlightTask-switch

### PR DESCRIPTION
**Pull request by Stifael**
This PR follows the issue described here https://github.com/PX4/Firmware/issues/13293.

Not sure if https://github.com/PX4/Firmware/compare/px4/bugfix/pass-timestamp-into-flighttasks?expand=1#diff-fc77c3ef569029d45764664c75d4b0c1R790-R800 is correct, but given that it was `control_mode.auto before`, I suspect that this is correct.

### CODE REVIEW

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 1]
```diff
@@ -783,19 +783,35 @@ MulticopterPositionControl::start_flight_task()
 			// we want to be in this mode, reset the failure count
 			_task_failure_count = 0;
 		}
+	}
 
-	} else if (_control_mode.flag_control_auto_enabled) {
+	// Auto
+	const bool vehicle_in_auto_mode = (_vehicle_status.nav_state == vehicle_status_s::NAVIGATION_STATE_AUTO_RTGS
```
Comment (dagar) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: What's this fixing? Is there a nav state that doesn't fit into control mode?

**Comment by dagar** [comment]
What's the fix for a FlightTaskAuto activating before the position_setpoint_triplet is available?

**Comment by Stifael** [comment]
> What's the fix for a FlightTaskAuto activating before the position_setpoint_triplet is available? 

https://github.com/PX4/Firmware/pull/13526/files#diff-30c13a2a739574585e6a798c0b6e61beR143

Basically forces the vehicle to hover at current location.

File: src/modules/mc_pos_control/mc_pos_control_main.cpp (hunk 1 above)
Comment (Stifael) [reply src/modules/mc_pos_control/mc_pos_control_main.cpp]: It is just more consistent because all other switches are caused by the `vehicle_nav_state`

**Comment by dagar** [comment]
Ok, this still seems pretty horrible. Let's pursue a proper fix as soon as possible. Can we discuss on the dev call wednesday?

**Comment by Stifael** [comment]
> Can we discuss on the dev call wednesday?

sure

File: src/lib/flight_tasks/FlightTasks.hpp [hunk 2]
```diff
@@ -109,15 +110,15 @@ class FlightTasks
 	 * Switch to the next task in the available list (for testing)
 	 * @return 0 on success, <0 on error
 	 */
-	FlightTaskError switchTask() { return switchTask(static_cast<int>(_current_task.index) + 1); }
+	FlightTaskError switchTask() { return switchTask(static_cast<int>(_current_task.index) + 1); };
```
Comment (dagar) [inline src/lib/flight_tasks/FlightTasks.hpp]:
```suggestion
	FlightTaskError switchTask() { return switchTask(static_cast<int>(_current_task.index) + 1); }
```

Not necessary.

File: src/lib/flight_tasks/FlightTasks.hpp [hunk 3]
```diff
@@ -109,15 +110,15 @@ class FlightTasks
...
 
 	/**
 	 * Switch to a specific task (for normal usage)
 	 * @param task index to switch to
 	 * @return 0 on success, <0 on error
 	 */
-	FlightTaskError switchTask(FlightTaskIndex new_task_index);
-	FlightTaskError switchTask(int new_task_index);
+	FlightTaskError switchTask(FlightTaskIndex new_task_index, const hrt_abstime time_stamp_mode_switch = 0);
```
Comment (dagar) [inline src/lib/flight_tasks/FlightTasks.hpp]:
```suggestion
	FlightTaskError switchTask(FlightTaskIndex new_task_index, const hrt_abstime& time_stamp_mode_switch = 0);
```

File: src/lib/flight_tasks/FlightTasks.hpp [hunk 4]
```diff
@@ -109,15 +110,15 @@ class FlightTasks
...
+	FlightTaskError switchTask(int new_task_index, const hrt_abstime time_stamp_mode_switch = 0);
```
Comment (dagar) [inline src/lib/flight_tasks/FlightTasks.hpp]:
```suggestion
	FlightTaskError switchTask(int new_task_index, const hrt_abstime& time_stamp_mode_switch = 0);
```

File: src/lib/flight_tasks/FlightTasks.cpp [hunk 5]
```diff
@@ -56,7 +56,7 @@ const landing_gear_s FlightTasks::getGear()
 	}
 }
 
-FlightTaskError FlightTasks::switchTask(FlightTaskIndex new_task_index)
+FlightTaskError FlightTasks::switchTask(FlightTaskIndex new_task_index, const hrt_abstime time_stamp_mode_switch)
```
Comment (dagar) [inline src/lib/flight_tasks/FlightTasks.cpp]:
```suggestion
FlightTaskError FlightTasks::switchTask(FlightTaskIndex new_task_index, const hrt_abstime& time_stamp_mode_switch)
```

File: src/lib/flight_tasks/FlightTasks.cpp [hunk 6]
```diff
@@ -87,7 +90,7 @@ FlightTaskError FlightTasks::switchTask(FlightTaskIndex new_task_index)
 	return FlightTaskError::NoError;
 }
 
-FlightTaskError FlightTasks::switchTask(int new_task_index)
+FlightTaskError FlightTasks::switchTask(int new_task_index, const hrt_abstime time_stamp_mode_switch)
```
Comment (dagar) [inline src/lib/flight_tasks/FlightTasks.cpp]:
```suggestion
FlightTaskError FlightTasks::switchTask(int new_task_index, const hrt_abstime& time_stamp_mode_switch)
```

File: src/lib/flight_tasks/tasks/FlightTask/FlightTask.hpp [hunk 7]
```diff
@@ -157,6 +157,11 @@ class FlightTask : public ModuleParams
 		updateParams();
 	}
 
+	/**
+	 *  Set the timestamp of the message that caused a FlightTasks-switch
+	 */
+	void setModeSwitchTimeStamp(const hrt_abstime t_stamp_mode_switch) {_time_stamp_mode_switch_requested = t_stamp_mode_switch;}
```
Comment (dagar) [inline src/lib/flight_tasks/tasks/FlightTask/FlightTask.hpp]:
```suggestion
	void setModeSwitchTimeStamp(const hrt_abstime& t_stamp_mode_switch) { _time_stamp_mode_switch_requested = t_stamp_mode_switch; }
```

**Comment by Stifael** [comment]
@MaEtUgR as discussed in the issue https://github.com/PX4/Firmware/issues/13293, this change is considered as a hotfix to prevent further crashes. However, I think but have not tested yet, the changes of this PR might not be necessary anymore with the changes of https://github.com/PX4/Firmware/pull/13534. 

> If we want to check for the timestamps this way please let us directly use the existing _time_stamp_activate

See https://github.com/PX4/Firmware/issues/13293#issuecomment-555518282. Assuming the decision was/is to use timestamp to check for valid triplets, then we cannot use `_time_stamp_activate` because that timestamp does not give us any information about when the commander triggered a state change.

>  every possible mode change to initialize the flight task state and then compare to timestamps which might but not necessarily are coming from the navigator

that is not true. These are two distinct timestamps: one that captures the timestamp of the commander state transition and one that represents the timestamp when flighttask was triggered. I would not mix them, because otherwise you would introduce inconsistency

**Comment by Stifael** [comment]
@MaEtUgR additional context: https://github.com/PX4/Firmware/issues/13293#issuecomment-547813644

**Comment by jorge789** [comment]
**Tested on PixRacer V4:**
**Modes Tested**

Position Mode: Good.
Altitude Mode: Good.
Stabilized Mode: Good.
Mission Plan Mode (Automated): Good.
RTL: Good.

**- Procedure**
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode proceed to switch to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoin activate RTL and see landing behaviour.

**Notes:**
No issues noted, good flight in general.

**Log:**  https://review.px4.io/plot_app?log=080cbdf5-b0e4-4c7e-8b78-d535e2e87d8c

**Comment by Junkim3DR** [comment]
### Tested on NXP FMUK66 v3

**Modes Tested**
Position Mode: Good.
Altitude Mode: Good.
Stabilized Mode: Good.
Mission Plan Mode (Automated): Good.
RTL: Good.

**- Procedure**
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode proceed to switch to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint activate RTL and see landing behaviour.

**Notes:**
No issues noted, good flight in general.

**Log:**  https://review.px4.io/plot_app?log=f1ad1779-5eb7-47d0-9492-acf0e775c8b1

**Comment by Tony3dr** [comment]
**Tested on Pixhawk 4 V5**
**Modes Tested**
Position Mode: Good.
Altitude Mode: Good.
Stabilized Mode: Good.
Mission Plan Mode (Automated): Good.
RTL: Good.

- Procedure
Arm and Takeoff in position mode, after flying for approximately one minute, switched to altitude then stabilized mode proceed to switch to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint activate RTL, vehicle landed smooth and disarmed. 

Notes:
No issues noted good flight in general.

Log: https://review.px4.io/plot_app?log=140453c2-6209-44ef-8300-75607cfe1ac8&fbclid=IwAR1RAzJhsxvxJPVVJqqtLBktvxJfZxEITW3HJvXkZvBBm4HPmA3vxlXGCbo

File: src/lib/flight_tasks/FlightTasks.hpp (hunk 4 above)
Comment (Stifael) [reply src/lib/flight_tasks/FlightTasks.hpp]: pass by reference

**Comment by MaEtUgR** [comment]
We just discussed at the end of the dev call that it would be desirable to have a sequential execution of navigator, flight task, position control which would automatically solve this synchronization problem. Since the flight task library was originally designed to work that way it seems desirable.

**Comment by TSC21** [comment]
@Stifael @MaEtUgR is this PR going to be dropped, based on the latest discussions and formed workgroup around the rearchitecture of the Navigator+FlightTasks?

**Pull request closed without merging**